pgrx/src/fcall.rs

```diff
@@ -0,0 +1,143 @@
+//LICENSE Portions Copyright 2019-2021 ZomboDB, LLC.
+//LICENSE
+//LICENSE Portions Copyright 2021-2023 Technology Concepts & Design, Inc.
+//LICENSE
+//LICENSE Portions Copyright 2023-2023 PgCentral Foundation, Inc. <contact@pgcentral.org>
+//LICENSE
+//LICENSE All rights reserved.
+//LICENSE
+//LICENSE Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+use crate::{
+    direct_function_call, list::PgList, pg_sys, pg_sys::AsPgCStr, Array, FromDatum, IntoDatum,
+};
+
+pub trait FCallArg {
+    fn as_datum(&self) -> Option<pg_sys::Datum>;
+    fn type_oid(&self) -> pg_sys::Oid;
+}
+
+impl<T: IntoDatum + Clone> FCallArg for Option<T> {
+    fn as_datum(&self) -> Option<pg_sys::Datum> {
+        // TODO:  would prefer not to need `Clone`, but that requires changes to `IntoDatum`
+        self.as_ref().map(|v| Clone::clone(v).into_datum()).flatten()
+    }
+
+    fn type_oid(&self) -> pg_sys::Oid {
+        T::type_oid()
+    }
+}
+
+// TODO:  This should return `Result<Option<T>, FCallError>`
+//        We'll want to trap the common errors like ERRCODE_UNDEFINED_FUNCTION and ERRCODE_AMBIGUOUS_FUNCTION
+//        along with making some of our own error types
+pub fn fcall<T: FromDatum>(fname: &str, args: &[&dyn FCallArg]) -> Option<T> {
```

Comment:
We recently had an interesting report that returning Results leads to worse UX in some cases than panicking, but I am not sure exactly what the context was. Not actionable yet.

Reply:
However:
```suggestion
// TODO:  This should return `Result<Option<T>, FCallError>`
//        We'll want to trap the common errors like ERRCODE_UNDEFINED_FUNCTION and ERRCODE_AMBIGUOUS_FUNCTION
//        along with making some of our own error types
#[cfg(not(feature = "pg11"))]
pub fn fcall<T: FromDatum>(fname: &str, args: &[&dyn FCallArg]) -> Option<T> {
```

Reply:
> We recently had an interesting report that returning Results leads to worse UX in some cases than panicking, but I am not sure exactly what the context was. Not actionable yet.

I saw that in passing. If it’s not reporting nicely that’s our fault, not specifically the fault of Result.